pkg/controller/containersnapshot/containersnapshot_controller.go

```diff
@@ -0,0 +1,411 @@
+		client:                mgr.GetClient(),
+		scheme:                mgr.GetScheme(),
+		workerImage:           os.Getenv(envKeyWorkerImage),
+		workerImagePullSecret: os.Getenv(envKeyWorkerImagePullSecret),
+		workerServiceAccount:  os.Getenv(envKeyWorkerServiceAccount),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("containersnapshot-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource ContainerSnapshot
+	err = c.Watch(&source.Kind{Type: &atomv1alpha1.ContainerSnapshot{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to secondary resource Pods and requeue the owner ContainerSnapshot
+	err = c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &atomv1alpha1.ContainerSnapshot{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileContainerSnapshot implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileContainerSnapshot{}
+
+// ReconcileContainerSnapshot reconciles a ContainerSnapshot object
+type ReconcileContainerSnapshot struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client                client.Client
+	scheme                *runtime.Scheme
+	workerImage           string
+	workerImagePullSecret string
+	workerServiceAccount  string
+}
+
+// Reconcile reads that state of the cluster for a ContainerSnapshot object and makes changes based on the state read
+// and what is in the ContainerSnapshot.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileContainerSnapshot) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ContainerSnapshot")
+
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+
+	// Fetch the ContainerSnapshot instance
+	instance := &atomv1alpha1.ContainerSnapshot{}
+	err := r.client.Get(ctx, request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	if !instance.DeletionTimestamp.IsZero() {
```

Comment:
`DeletionTimestamp` may be nil

Reply:
it is checked in `IsZero()` method

Reply:
`nil.IsZero()` ?

Reply:
Actually something like `(*time.Time)(nil).IsZero()` is not guarrenteed to panic, see https://groups.google.com/forum/#!msg/golang-nuts/wcrZ3P1zeAk/WI88iQgFMvwJ . But `time.Time.IsZero()` does not check nil pointers and it will panic when called on nil pointers.

`DeletionTimestamp` is defined as `k8s.io/apimachinery/pkg/apis/meta/v1.Time` which wraps `time.Time` and checkes nil pointers.

Reply:
ok, didn't notice it's wrapped